src/transformers/models/bert/modeling_flax_bert.py

```diff
@@ -890,63 +909,115 @@ def __call__(
-                return outputs
             elif past_key_values is not None and not return_dict:
                 outputs, past_key_values = outputs
                 outputs = outputs[:1] + (unfreeze(past_key_values["cache"]),) + outputs[1:]
-
         else:
-            outputs = self.module.apply(
-                inputs,
-                jnp.array(input_ids, dtype="i4"),
-                jnp.array(attention_mask, dtype="i4"),
-                token_type_ids=jnp.array(token_type_ids, dtype="i4"),
-                position_ids=jnp.array(position_ids, dtype="i4"),
-                head_mask=jnp.array(head_mask, dtype="i4"),
-                deterministic=not train,
-                output_attentions=output_attentions,
-                output_hidden_states=output_hidden_states,
-                return_dict=return_dict,
-                rngs=rngs,
-            )
+            if params is not None:
+                raise ValueError(
+                    f"`params` should be passed when {self.__class__.__name__} is bounded to a scope "
+                    "(i.e. when calling this module with `apply`)"
+                )
+            if past_key_values is not None:
+                raise ValueError(
+                    f"`past_key_values` should passed when {self.__class__.__name__} is bounded to a scope."
+                    f"`Instead pass `past_key_values` to `apply` through the `cache` collection."
+                )
+            if self.config.add_cross_attention:
+                outputs = self.module(
+                    jnp.array(input_ids, dtype="i4"),
+                    jnp.array(attention_mask, dtype="i4"),
+                    token_type_ids=jnp.array(token_type_ids, dtype="i4"),
+                    position_ids=jnp.array(position_ids, dtype="i4"),
+                    head_mask=jnp.array(head_mask, dtype="i4"),
+                    encoder_hidden_states=encoder_hidden_states,
+                    encoder_attention_mask=encoder_attention_mask,
+                    deterministic=not train,
+                    output_attentions=output_attentions,
+                    output_hidden_states=output_hidden_states,
+                    return_dict=return_dict,
+                )
+            else:
+                outputs = self.module(
+                    jnp.array(input_ids, dtype="i4"),
+                    jnp.array(attention_mask, dtype="i4"),
+                    token_type_ids=jnp.array(token_type_ids, dtype="i4"),
+                    position_ids=jnp.array(position_ids, dtype="i4"),
+                    head_mask=jnp.array(head_mask, dtype="i4"),
+                    deterministic=not train,
+                    output_attentions=output_attentions,
+                    output_hidden_states=output_hidden_states,
+                    return_dict=return_dict,
+                )
+
+            # remove the cache from the output, the user should retrieve it from the cache collection
```

Comment:
Cool! So the behaviour of the cache is unaffected if `_do_init=True` (i.e. stateful), but now needs to be handled in the cache collection if `_do_init=False` (i.e. stateless)